UselessReturnFixer - Add

SpacePossum opened the pull request:
fabbot is wrong on this one, but a fix(er) for this is already PR'ed ;)

GrahamCampbell commented on README.rst:
```diff
@@ -659,6 +659,11 @@ Choose from the list of available fixers:
                         $strict param. Warning! This
                         could change code behavior.
 
+* **useless_return** [contrib]
```
This should be symfony, right?

SpacePossum replied:
not sure, happy to move though :)

SpacePossum replied:
ping @keradus, should I move it?

keradus replied:
@SpacePossum I wanted to think about it but the fixer doesn't work ;)

SpacePossum replied:
there are about 6 cases on SF 2.3

   1) Bundle/WebProfilerBundle/Tests/Controller/ProfilerControllerTest.php (useless_return)
   2) Component/Intl/Data/Generator/LanguageDataGenerator.php (useless_return)
   3) Component/Intl/Data/Generator/CurrencyDataGenerator.php (useless_return)
   4) Component/Form/Tests/Extension/Validator/Type/BaseValidatorExtensionTest.php (useless_return)
   5) Component/Finder/Adapter/AbstractFindAdapter.php (useless_return)
   6) Component/Serializer/Tests/Normalizer/GetSetMethodNormalizerTest.php (useless_return)

keradus replied:
So standard procedure - send a PR to fix them.
Looking on the scale I think it will be merged, so Symfony level indeed.

SpacePossum replied:
check, done (https://github.com/symfony/symfony/pull/17249)
lets see what comes back before I move the fixer.

SpacePossum replied:
I suggest leaving this Fixer at `contrib`, we can always move it to Symfony level later when they settled the return null vs. return void issue.

keradus replied:
fine with me

GrahamCampbell commented on Symfony/CS/Tests/Fixtures/Integration/empty_return,useless_return,whitespacy_lines,extra_empty_lines.test:
```diff
@@ -0,0 +1,43 @@
+        if ($a) {
+            return 1;
+        }
+
```
this line should be removed, right?

GrahamCampbell replied:
if the return is on it's own line, it would make sense to clear the whole line inc the newline char, right?

SpacePossum replied:
The `extra_empty_lines` should pick this one up, we should add more config options to that one.
Which reminds me, this fixer should also run before the `return` fixer to prevent that one from adding an empty line.

GrahamCampbell replied:
Yeh ;)

keradus commented on Symfony/CS/Tests/FixerTest.php:
```diff
@@ -209,6 +209,9 @@ public function getFixersPriorityCases()
             array($fixers['class_definition'], $fixers['trailing_spaces']), // tested also in: class_definition,trailing_spaces.test
+            array($fixers['empty_return'], $fixers['useless_return']), // tested also in: empty_return,useless_return,whitespacy_lines,extra_empty_lines.test
+            array($fixers['useless_return'], $fixers['whitespacy_lines']), // tested also in: empty_return,useless_return,whitespacy_lines,extra_empty_lines.test
+            array($fixers['useless_return'], $fixers['extra_empty_lines']), // tested also in: empty_return,useless_return,whitespacy_lines,extra_empty_lines.test
```
Please, also add integration test for only pair of fixers - one could run only 2 of them.

SpacePossum replied:
not sure what such tests would add, the integration test already proves that changing any of the fixers priorities will cause troubles

keradus replied:
Will be extra helpfull during any refactoring. It's always great to have more detailed tests as well, not only "big" one.

SpacePossum replied:
check, added more tests :)

keradus commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,101 @@
+                    break;
+                }
+                $content .= $tokens[$i]->getContent();
+                $tokens[$i]->clear();
```
You need to merge 2 T_WHITESPACE being placed one after another. Where could be the situation with more than 2 ? I think this loop is not needed.

keradus commented on Symfony/CS/Tests/Fixer/Contrib/UselessReturnFixerTest.php:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @internal
+ */
+class UselessReturnFixerTest extends AbstractFixerTestBase
```
final

keradus commented on Symfony/CS/Tests/Fixer/Contrib/UselessReturnFixerTest.php:
```diff
@@ -0,0 +1,107 @@
+        $this->makeTest($expected, $input);
+    }
+
+    public function provideFixCases()
```
looking at the implementation it should work, but please add case with nested functions.

SpacePossum replied:
thanks and done :)

keradus commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,101 @@
+            }
+
+            $functionOpen = $tokens->getNextTokenOfKind($index, array('{'));
+            $this->fixFunction($tokens, $functionOpen, $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $functionOpen));
```
1. iterate backward all tokens
2. for each T_FUNCTION iterate forward to find it's end
3. for each T_FUNCTION iterate backward from end to start

This means that for each function you are iterating through all its tokens... 3 times!
And or any 1st level nested function - 3x2, 2nd level nested function - 3x3 times and so on...

Here, you are not adding any tokens, only removing/merging. Then, you could Iterate forward on main loop and have no need to iterate few times over same tokens.

SpacePossum replied:
please have another look, I took out some unneeded iterations

keradus commented on README.rst:
```diff
@@ -659,6 +659,11 @@ Choose from the list of available fixers:
                         could change code behavior.
 
+* **useless_return** [contrib]
+                        These should not be an empty
```
`There should not (...)` ?

gharlan commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,96 @@
+     */
+    public function getDescription()
+    {
+        return 'There should not be an empty return statement at the end of a function.';
```
why only at end of function, why not at end of file?

SpacePossum replied:
I think that would make the fixer `risky`, for example:
```php
<?php
class Test {

    public function myTestFunction($a)
    {
        $b = $a;
        if ($a) {
            $b = include 'includeTest.php';
        }
        return $b;
    }
}

$test = new Test();
$b = $test->myTestFunction(true);
var_dump($b);
die;
```

`includeTest.php`

```php
<?php

if (time() < 0) {
    return 1;
}
return;
```

sure it isn't the most clean code IMHO, but it is valid and the fixer would break it.

gharlan replied:
I thought that default return value of `include` is `null`. But it is wrong, default return value is `1`.

So I agree, would be risky and we should not change it.

keradus replied:
Sadly, this is taken from perl. It is a confirmation that the module has been loaded :(

Please do not mess with file-level return statement in this one fixer. For sure it should not be handled in same fixer (non-risky + risky, no need to mark first part as risky as well).
Also, I don't think that it is needed at all. If one put empty return on non-function code then we could assume that he want to change that default `1` value ;)

gharlan replied:
Yeah, we all agree here: nothing to change in this fixer and no need for another fixer.

keradus commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        // merge whitespace tokens if needed
+        if ($tokens[$start - 1]->isWhitespace() && $tokens[$semiColonIndex + 1]->isWhitespace()) {
```
what if between `$start - 1` and `$semiColonIndex + 1` is that ommited comment ?

SpacePossum replied:
good point, fixed :)

SpacePossum commented on Symfony/CS/Tests/Fixtures/Integration/empty_return,useless_return,whitespacy_lines,extra_empty_lines.test:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+--EXPECT--
```
@keradus I should swap the expected and input part right?

keradus replied:
when #1605 will be merged

SpacePossum replied:
:+1:  
which is RTM, so let me switch those test around :)
(expecting tests to fail)

keradus commented:
https://github.com/SpacePossum/PHP-CS-Fixer/pull/6

SpacePossum commented:
Thanks @keradus, picked in this branched, rebased and clean upped a bit. Lets see how the test do...
In the mean time could you think about https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1591#discussion_r47430303?

GrahamCampbell commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,98 @@
+
+            $this->removeReturnStatement($tokens, $index, $nextAt);
+
+            return;
```
Isn't writing `break;` more conventional here. I think `return;` is a little confusing, especially if you missread the closing braces.

GrahamCampbell commented on Symfony/CS/Tests/Fixer/Symfony/EmptyReturnFixerTest.php:
```diff
@@ -30,7 +30,6 @@ public function provideExamples()
     {
         return array(
             // check correct statements aren't changed
-            array('<?php return;'),
```
?

keradus replied:
The test is duplicated few lines later

keradus commented:
You squash it, could you point out the fix?

keradus commented:
Oh, there are no fix yet ;)

SpacePossum commented:
>You squash it, could you point out the fix?

not squashed (yet ;) )
https://github.com/SpacePossum/PHP-CS-Fixer/commit/102eec2d7d06e72e9c4a1ef6469350d46390a1ba

keradus commented:
https://github.com/SpacePossum/PHP-CS-Fixer/pull/7 ?

keradus commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private function removeReturnStatement(Tokens $tokens, $start, $semiColonIndex)
```
`$semicolonIndex`

phpdoc (at least typehinting)

keradus commented:
Please add:
```php
            array(
                '<?php
    function foo () { }',
            ),
            array(
                '<?php
    function foo () {
        '.'
    }',
                '<?php
    function foo () {
        return ;;
    }',
            ),
```

SpacePossum commented:
Thanks @keradus :)
You second test case is invalid, however it shows that the fixer should be ran after the `duplicate_semicolon`. 
I'll do some updating for this soon.

keradus commented:
> You second test case is invalid

Why is that ?

SpacePossum commented:
yes, meant 3rd, I was counting from 0 I guess 0.o

keradus commented:
:D

keradus commented:
Please rebase and move integration tests into `priority` subdir

keradus commented:
Please @SpacePossum squash my commits into 1

SpacePossum commented:
I've squashed your commits.. but I'm sorry about the result...

gharlan commented:
from https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1672#issuecomment-171590838:

> the useless return fixer should be able to deal with multiple useless returns after each order (current missing feature)

this should be fixed

SpacePossum commented:
appveyor is down "Failed to provision build worker virtual machine in Google Compute Engine cloud. Try restarting build later." .. will try again later

I'll work on the multiples soon... would it stop merging though?

gharlan commented:
> I'll work on the multiples soon... would it stop merging though?

IMO yes. A fixer should not change anything in a second run.

SpacePossum commented:
>IMO yes. A fixer should not change anything in a second run.

didn't thought about that, agreed, I'll rework this :+1:

keradus commented on Symfony/CS/Fixer/Contrib/UselessReturnFixer.php:
```diff
@@ -0,0 +1,108 @@
+            $tokens[$index]->clear();
+        }
+
+        // merge whitespace tokens if needed
```
it is the next place when I see this... please follow DRY

SpacePossum replied:
This is done by a lot fixers, all in their own ways. I'm not sure where to refactor to, however I feel this fixer is a bit of a mess at the moment.

keradus replied:
Feel free to clean it up ;)

SpacePossum replied:
>I'm not sure where to refactor to

;)

keradus commented:
btw, please rebase ;)

SpacePossum commented:
>btw, please rebase ;)

:+1: done :)

keradus commented:
:+1:

before merge please squash commits (for some reason my one already lost almost all content?) - squash all into one (so even my one) ;)
then, please add fixer to `.php_cs` file

SpacePossum commented:
>for some reason my one already lost almost all content?

yea sorry about that, I've no idea how I kept your commit but lost a part of the content :(

I've added the fixer the `.php_cs`, squashed all commits into one.

keradus commented:
very well
